pkg/api/osa/vlabs/types.go

```diff
@@ -0,0 +1,133 @@
+	Location   string                `json:"location,omitempty"`
+	Name       string                `json:"name,omitempty"`
+	Plan       *ResourcePurchasePlan `json:"plan,omitempty"`
+	Tags       map[string]string     `json:"tags,omitempty"`
+	Type       string                `json:"type,omitempty"`
+	Properties *Properties           `json:"properties,omitempty"`
+}
+
+// ResourcePurchasePlan defines the resource plan as required by ARM for billing
+// purposes.
+type ResourcePurchasePlan struct {
+	Name          string `json:"name,omitempty"`
+	Product       string `json:"product,omitempty"`
+	PromotionCode string `json:"promotionCode,omitempty"`
+	Publisher     string `json:"publisher,omitempty"`
+}
+
+// Properties represents the cluster definition.
+type Properties struct {
+	// ProvisioningState (out): current state of the OSA resource.
+	ProvisioningState ProvisioningState `json:"provisioningState,omitempty"`
+
+	// OpenShiftVersion (in): OpenShift version to be created/updated, e.g.
+	// `v3.10`.
+	OpenShiftVersion string `json:"openShiftVersion,omitempty"`
+
+	// PublicHostname (in,optional): Optional user-specified FQDN for OpenShift
+	// API server.  If specified, after OSA cluster creation, user must create a
+	// PublicHostname CNAME record forwarding to the returned FQDN value.
+	PublicHostname string `json:"publicHostname,omitempty"`
+
+	// FQDN (out): Auto-allocated FQDN for OpenShift API server.
+	FQDN string `json:"fqdn,omitempty"`
+
+	// RouterProfiles (in,optional/out): Configuration for OpenShift router(s).
+	RouterProfiles []RouterProfile `json:"routerProfiles,omitempty"`
+
+	// AgentPoolProfiles (in): configuration of OpenShift cluster VMs.
+	AgentPoolProfiles []AgentPoolProfile `json:"agentPoolProfiles,omitempty"`
+
+	// TODO: is this compatible with MSI?
+	// ServicePrincipalProfile (in): Service principal for OpenShift cluster.
+	ServicePrincipalProfile ServicePrincipalProfile `json:"servicePrincipalProfile,omitempty"`
+}
+
+// ProvisioningState represents the current state of the OSA resource.
+type ProvisioningState string
+
+const (
+	// Creating means the OSA resource is being created.
+	Creating ProvisioningState = "Creating"
+	// Updating means the existing OSA resource is being updated.
+	Updating ProvisioningState = "Updating"
+	// Failed means the OSA resource is in failed state.
+	Failed ProvisioningState = "Failed"
+	// Succeeded means the last create/update succeeded.
+	Succeeded ProvisioningState = "Succeeded"
+	// Deleting means the OSA resource is being deleted.
+	Deleting ProvisioningState = "Deleting"
+	// Migrating means the OSA resource is being migrated from one subscription
+	// or resource group to another.
+	Migrating ProvisioningState = "Migrating"
+	// Upgrading means the existing OAS resource is being upgraded.
+	Upgrading ProvisioningState = "Upgrading"
+)
+
+// RouterProfile represents an OpenShift router.
+type RouterProfile struct {
+	Name string `json:"name,omitempty"`
```

Comment:
Is the name for each `RouterProfile ` for easy referencing?

Reply:
yes; the name "default" might have slightly special powers as well.